### ISSUE TITLE
Dataloss in Libreview

### ISSUE THREAD

**Issue by VanessaMGoncalves** [j-kaltes/Juggluco#14]
After using the App sometimes the data is not fully saved in the Cloud (libreview). Any solution to fix this problem?

**Comment by j-kaltes** [comment]
Juggluco doesn't save everything immediately. It waits at least 20 minutes after the previous upload and you should switch away from Juggluco by turning off the screen or going to another app after having used Juggluco.

Show me the saved history values of Juggluco (Left middle menu->Export->History) and the saved values from Libreview (Download glucose data) and I will look at it.

**Comment by VanessaMGoncalves** [comment]
For example: 

Juggluco data in mg/dL:

XX0ATRJ469R	1671	1678354030	2023-03-09-10:27:10	1	8350	140
XX0ATRJ469R	1672	1678354332	2023-03-09-10:32:12	1	8355	136
XX0ATRJ469R	1673	1678354632	2023-03-09-10:37:12	1	8360	130
XX0ATRJ469R	1674	1678354932	2023-03-09-10:42:12	1	8365	118
XX0ATRJ469R	1675	1678355232	2023-03-09-10:47:12	1	8370	109
XX0ATRJ469R	1676	1678355532	2023-03-09-10:52:12	1	8375	104
XX0ATRJ469R	1677	1678355832	2023-03-09-10:57:12	1	8380	95
XX0ATRJ469R	1678	1678356132	2023-03-09-11:02:12	1	8385	87
XX0ATRJ469R	1679	1678356432	2023-03-09-11:07:12	1	8390	87
XX0ATRJ469R	1680	1678356732	2023-03-09-11:12:12	1	8395	87
XX0ATRJ469R	1681	1678357032	2023-03-09-11:17:12	1	8400	89
XX0ATRJ469R	1682	1678357332	2023-03-09-11:22:12	1	8405	91
XX0ATRJ469R	1683	1678357632	2023-03-09-11:27:12	1	8410	93
XX0ATRJ469R	1684	1678357932	2023-03-09-11:32:12	1	8415	96
XX0ATRJ469R	1685	1678358232	2023-03-09-11:37:12	1	8420	99
XX0ATRJ469R	1686	1678358532	2023-03-09-11:42:12	1	8425	99
XX0ATRJ469R	1687	1678358832	2023-03-09-11:47:12	1	8430	98
XX0ATRJ469R	1688	1678359132	2023-03-09-11:52:12	1	8435	97
XX0ATRJ469R	1689	1678359432	2023-03-09-11:57:12	1	8440	96
XX0ATRJ469R	1690	1678359732	2023-03-09-12:02:12	1	8445	94
XX0ATRJ469R	1691	1678360032	2023-03-09-12:07:12	1	8450	93
XX0ATRJ469R	1692	1678360332	2023-03-09-12:12:12	1	8455	95
XX0ATRJ469R	1693	1678360632	2023-03-09-12:17:12	1	8460	98
XX0ATRJ469R	1694	1678360932	2023-03-09-12:22:12	1	8465	105
XX0ATRJ469R	1695	1678361232	2023-03-09-12:27:12	1	8470	114
XX0ATRJ469R	1696	1678361532	2023-03-09-12:32:12	1	8475	123
XX0ATRJ469R	1697	1678361832	2023-03-09-12:37:12	1	8480	134
XX0ATRJ469R	1698	1678362132	2023-03-09-12:42:12	1	8485	139
XX0ATRJ469R	1699	1678362432	2023-03-09-12:47:12	1	8490	132
XX0ATRJ469R	1700	1678362732	2023-03-09-12:52:12	1	8495	118
XX0ATRJ469R	1701	1678363032	2023-03-09-12:57:12	1	8500	106
XX0ATRJ469R	1702	1678363332	2023-03-09-13:02:12	1	8505	98
XX0ATRJ469R	1703	1678363632	2023-03-09-13:07:12	1	8510	91
XX0ATRJ469R	1704	1678363932	2023-03-09-13:12:12	1	8515	89
XX0ATRJ469R	1705	1678364232	2023-03-09-13:17:12	1	8520	94
XX0ATRJ469R	1706	1678364530	2023-03-09-13:22:10	1	8525	102
XX0ATRJ469R	1707	1678364831	2023-03-09-13:27:11	1	8530	110
XX0ATRJ469R	1708	1678365130	2023-03-09-13:32:10	1	8535	116
XX0ATRJ469R	1709	1678365430	2023-03-09-13:37:10	1	8540	119
XX0ATRJ469R	1710	1678365730	2023-03-09-13:42:10	1	8545	119
XX0ATRJ469R	1711	        1678366031	2023-03-09-13:47:11	1	8550	121
XX0ATRJ469R	1712	1678366330	2023-03-09-13:52:10	1	8555	126
XX0ATRJ469R	1713	1678366631	2023-03-09-13:57:11	1	8560	130
XX0ATRJ469R	1714	1678366929	2023-03-09-14:02:09	1	8565	132
XX0ATRJ469R	1715	1678367231	2023-03-09-14:07:11	1	8570	133


Libreview (cloud data) (I added a last column with the results in mg/dL:


FreeStyle | Libre | 3 | 215f1df4-69a5-4026-b579-b673162cdb7c | 09/03/2023 | 10:25 | 1 | 8.6 | 155
FreeStyle | Libre | 3 | 215f1df4-69a5-4026-b579-b673162cdb7c | 09/03/2023 | 10:41 | 1 | 6.8 | 123
FreeStyle | Libre | 3 | 215f1df4-69a5-4026-b579-b673162cdb7c | 09/03/2023 | 10:44 | 1 | 6.5 | 117
FreeStyle | Libre | 3 | 215f1df4-69a5-4026-b579-b673162cdb7c | 09/03/2023 | 10:47 | 1 | 6.3 | 113
FreeStyle | Libre | 3 | 215f1df4-69a5-4026-b579-b673162cdb7c | 09/03/2023 | 13:40 | 1 | 6.7 | 121
FreeStyle | Libre | 3 | 215f1df4-69a5-4026-b579-b673162cdb7c | 09/03/2023 | 13:45 | 1 | 6.4 | 115
FreeStyle | Libre | 3 | 215f1df4-69a5-4026-b579-b673162cdb7c | 09/03/2023 | 13:56 | 1 | 7.5 | 135
FreeStyle | Libre | 3 | 215f1df4-69a5-4026-b579-b673162cdb7c | 09/03/2023 | 14:00 | 1 | 7.7 | 139
FreeStyle | Libre | 3 | 215f1df4-69a5-4026-b579-b673162cdb7c | 09/03/2023 | 14:04 | 1 | 7.7 | 139
FreeStyle | Libre | 3 | 215f1df4-69a5-4026-b579-b673162cdb7c | 09/03/2023 | 14:10 | 1 | 7.7 | 139


Also here you can notice that the time stamps between the two sources are not exactly the same (as mentioned in Issue #12).

**Comment by j-kaltes** [comment]
Can't you give me the files with the raw values? There have to be much more values.
EDIT:
For example send them to jaapkorthalsaltes@gmail.com or attach them with "Attach files" below the editing field.

You need the record type 0, you show the record type 1 that are the individual stream values that are shown as a unfilled circles in Libreview.
Record type 1 are for example:
FreeStyle Libre 3,d5b2ee2b-1f2d-4376-8a2c-683519dd0df3,20-03-2023 12:03,1,,7.1,,,,,,,,,,,,,
FreeStyle Libre 3,d5b2ee2b-1f2d-4376-8a2c-683519dd0df3,20-03-2023 12:38,1,,6.6,,,,,,,,,,,,,
FreeStyle Libre 3,d5b2ee2b-1f2d-4376-8a2c-683519dd0df3,20-03-2023 13:03,1,,6.3,,,,,,,,,,,,,
Record type 0 are:
FreeStyle Libre 3,d5b2ee2b-1f2d-4376-8a2c-683519dd0df3,20-03-2023 12:01,0,7.2,,,,,,,,,,,,,,
FreeStyle Libre 3,d5b2ee2b-1f2d-4376-8a2c-683519dd0df3,20-03-2023 12:06,0,7.0,,,,,,,,,,,,,,
FreeStyle Libre 3,d5b2ee2b-1f2d-4376-8a2c-683519dd0df3,20-03-2023 12:11,0,6.9,,,,,,,,,,,,,,
FreeStyle Libre 3,d5b2ee2b-1f2d-4376-8a2c-683519dd0df3,20-03-2023 12:16,0,6.8,,,,,,,,,,,,,,
FreeStyle Libre 3,d5b2ee2b-1f2d-4376-8a2c-683519dd0df3,20-03-2023 12:21,0,6.7,,,,,,,,,,,,,,
FreeStyle Libre 3,d5b2ee2b-1f2d-4376-8a2c-683519dd0df3,20-03-2023 12:26,0,6.7,,,,,,,,,,,,,,
FreeStyle Libre 3,d5b2ee2b-1f2d-4376-8a2c-683519dd0df3,20-03-2023 12:31,0,6.7,,,,,,,,,,,,,,
FreeStyle Libre 3,d5b2ee2b-1f2d-4376-8a2c-683519dd0df3,20-03-2023 12:36,0,6.7,,,,,,,,,,,,,,
FreeStyle Libre 3,d5b2ee2b-1f2d-4376-8a2c-683519dd0df3,20-03-2023 12:41,0,6.7,,,,,,,,,,,,,,
FreeStyle Libre 3,d5b2ee2b-1f2d-4376-8a2c-683519dd0df3,20-03-2023 12:46,0,6.6,,,,,,,,,,,,,,
FreeStyle Libre 3,d5b2ee2b-1f2d-4376-8a2c-683519dd0df3,20-03-2023 12:51,0,6.6,,,,,,,,,,,,,,
FreeStyle Libre 3,d5b2ee2b-1f2d-4376-8a2c-683519dd0df3,20-03-2023 12:56,0,6.5,,,,,,,,,,,,,,
FreeStyle Libre 3,d5b2ee2b-1f2d-4376-8a2c-683519dd0df3,20-03-2023 13:01,0,6.4,,,,,,,,,,,,,,

**Comment by VanessaMGoncalves** [comment]
Thanks for the feedback. I sent the files to your e-mail since I can't share a txt file here!

But for the previous example I gave you, you are right! I should have selected type 0 and not type 1. Thanks for noticing!

**Comment by kubasaw** [comment]
I'm not sure if this is the right thread, but it seems like I have a similar issue. It happens when there's a break in readings from Libre over BLE. For example, last night, the app wasn’t receiving data from Libre for about two hours. Later, the connection came back, and the app started sending data to Libreview again. That seems fine.

The problem shows up when I notice such a break and scan the sensor with NFC. The data shows up in Juggluco, but it doesn’t get sent to the cloud anymore. I think the app works in a way that it saves the timestamp of the last sent reading and then only adds new data from that point on. Because of this, any data that "appears in the past" (after an NFC scan) doesn’t get sent to Libreview.

The original Libre app doesn’t have this issue – gaps from the last 8 hours are "filled in" after an NFC scan.

**Comment by j-kaltes** [comment]
Is the Libre 2 sensor connected to an arm64 phone or an arm32 wearOS watch? And if it is a watch, do you have checked _Bluetooth history_?

**Comment by kubasaw** [comment]
Sensor is connected to arm64 phone (motorola edge series).

As you can see on the screenshot, BLE communication issues occurs between 4:30 and 6:30. From 6:30 when app was able to read sensor via BLE, data connection to Libreview was restored. Before I scanned it at 8:30, there was also gap (4:30-6:30) in cyan graph and inside Libreview. After scanning, gap was filled in Juggluco, but not inside Libreview.

<img src="https://github.com/user-attachments/assets/505c16c9-045c-471c-ac6b-58e246c5c92d" width=500px>
<img src="https://github.com/user-attachments/assets/b5261ff6-9d63-4bd2-96e8-8cdf706bf3b4" width=500px>

Apart from that, thank you very much for your time and work put into the application!

**Comment by j-kaltes** [comment]
I have just tried, it and the scanned data are later added to Libreview. I disconnected the sensor at 12:33.
At 13:40, just before scanning:
![before3](https://github.com/user-attachments/assets/b389617f-d68f-4ec8-9441-b0f27932b142)
At 13:41, after scanning:
![after3](https://github.com/user-attachments/assets/c6fbb1e5-c57d-4c7f-a875-3badecba2233)

Did you switch to a new sensor in the meantime? Juggluco tries to behave like Librelink by only sending from one sensor at a time. When data is missing from the first sensor, Juggluco switches to sending from the second sensor and never returns to the first sensor.

**Comment by kubasaw** [comment]
I'm working with the same sensor all the time. In a case like the one you described, I get exactly the same behavior: if the sensor disconnects and I scan it with NFC, the missing data is sent to Libreview. The data was transmitted because the scan happened **immediately after** the period of no BLE connection. 

If the sensor reconnects over BLE without scanning (let’s say the disconnection lasts an hour), there will be a one-hour gap in the history (both in Juggluco and Libreview). If you wait another hour or two, data (transmitted via BLE) from the moment of reconnection (the end of the gap) will be sent. However, if you then perform an NFC scan, the one-hour gap will be filled in Juggluco but not in Libreview. 

The only quick way to fill that gap in Libreview is to go into the Libreview connection settings and change the `ChangeStart` field to a timestamp from before the BLE disconnection. 

To sum it up as simply as I can: If I lost BLE connection at 0:00 and regained it at 1:00, Juggluco starts sending data from 1:00 onwards (I suspect it saves the timestamp of the last transmitted reading at that point). By 7:00, I have gap from 0:00 to 1:00 and data from 1:00 to 7:00 in both LibreView and Juggluco (received over BLE), and the timestamp of the last transmitted reading inside Juggluco is set to 7:00. 

If I then scan the sensor with NFC, Juggluco fills the missing history gap from 0:00 to 1:00, but this data isn’t sent to LibreView because the timestamps of these readings are earlier than 7:00.

**Comment by j-kaltes** [comment]
Yes, it seemed to contain a bug. I uploaded version 9.0.12 without this problem:
https://www.juggluco.nl/Juggluco/download.html

I just tested it, by saving the data with "_Download Glucose Data_" in Libreview. After disconnecting between 16:48 and 1740 and then receiving two history values via bluetooth:
FreeStyle LibreLink,d85ea27b-d889-4d77-bf2b-7b7055ba45d8,09-12-2024 16:30,0,4.4,,,,,,,,,,,,,,
FreeStyle LibreLink,d85ea27b-d889-4d77-bf2b-7b7055ba45d8,09-12-2024 17:15,0,6.7,,,,,,,,,,,,,,
FreeStyle LibreLink,d85ea27b-d889-4d77-bf2b-7b7055ba45d8,09-12-2024 17:31,0,5.9,,,,,,,,,,,,,,
After scanning:
FreeStyle LibreLink,d85ea27b-d889-4d77-bf2b-7b7055ba45d8,09-12-2024 16:30,0,4.4,,,,,,,,,,,,,,
FreeStyle LibreLink,d85ea27b-d889-4d77-bf2b-7b7055ba45d8,09-12-2024 16:47,0,4.6,,,,,,,,,,,,,,
FreeStyle LibreLink,d85ea27b-d889-4d77-bf2b-7b7055ba45d8,09-12-2024 17:02,0,5.8,,,,,,,,,,,,,,
FreeStyle LibreLink,d85ea27b-d889-4d77-bf2b-7b7055ba45d8,09-12-2024 17:15,0,6.7,,,,,,,,,,,,,,
FreeStyle LibreLink,d85ea27b-d889-4d77-bf2b-7b7055ba45d8,09-12-2024 17:31,0,5.9,,,,,,,,,,,,,,

**Comment by kubasaw** [comment]
Today I have the same problem at night and I think it is totally fixed. Thank you again! You fixed it faster than I was able to think, where the bug may be located :)
